shared/reducers/waiting.tsx

```diff
@@ -1,71 +1,81 @@
-  let newErrors = state.errors
+const changeHelper = (
+  draftState: Container.Draft<Types.State>,
+  keys: string | Array<string>,
+  diff: 1 | -1,
+  error?: RPCError
+) => {
+  const counts = new Map(draftState.counts)
+  const errors = new Map(draftState.errors)
 
-  keys.forEach(k => {
-    const oldCount = newCounts.get(k, 0)
+  getKeys(keys).forEach(k => {
+    const oldCount = draftState.counts.get(k) || 0
     // going from 0 => 1, clear errors
     if (oldCount === 0 && diff === 1) {
-      newErrors = newErrors.set(k, null)
+      errors.delete(k)
     } else {
       if (error) {
-        newErrors = newErrors.set(k, error)
+        errors.set(k, error)
       }
     }
     const newCount = oldCount + diff
     if (newCount === 0) {
-      newCounts = newCounts.delete(k)
+      counts.delete(k)
     } else {
-      newCounts = newCounts.set(k, newCount)
+      counts.set(k, newCount)
     }
   })
 
-  const newState = state.merge({counts: newCounts, errors: newErrors})
-  debugWaiting && console.log('DebugWaiting:', keys, newState.toJS())
-  return newState
+  debugWaiting && console.log('DebugWaiting:', keys, draftState)
+  draftState.counts = counts
+  draftState.errors = errors
 }
 
-const initialState = Constants.makeState()
-function reducer(state: Types.State = initialState, action: Waiting.Actions): Types.State {
-  switch (action.type) {
-    case 'common:resetStore': {
-      // Keep the old values else the keys will be all off and confusing
-      const newState = initialState.merge(state)
-      debugWaiting && console.log('DebugWaiting:', '*resetStore*', newState.toJS())
-      return newState
-    }
-    case Waiting.decrementWaiting: {
-      const {key, error} = action.payload
-      return changeHelper(state, typeof key === 'string' ? [key] : key, -1, error)
-    }
-    case Waiting.incrementWaiting: {
-      const {key} = action.payload
-      return changeHelper(state, typeof key === 'string' ? [key] : key, 1)
-    }
-    case Waiting.clearWaiting: {
-      const {key} = action.payload
-      return state.merge({
-        counts: state.counts.deleteAll(typeof key === 'string' ? [key] : key),
-        errors: state.errors.deleteAll(typeof key === 'string' ? [key] : key),
-      })
-    }
-    case Waiting.batchChangeWaiting: {
-      const {changes} = action.payload
-      return changes.reduce((state, {key, increment, error}) => {
-        return changeHelper(state, typeof key === 'string' ? [key] : key, increment ? 1 : -1, error)
-      }, state)
-    }
-    default:
-      return state
-  }
+const initialState: Types.State = {
+  counts: new Map<string, number>(),
+  errors: new Map<string, RPCError>(),
 }
 
-export default reducer
+const getKeys = (k: string | Array<string>) => {
+  if (typeof k === 'string') return [k]
+  return k
+}
+
+export default (state: Types.State = initialState, action: Waiting.Actions): Types.State =>
+  Container.produce(state, (draftState: Container.Draft<Types.State>) => {
```

Comment:
Can we skip the type annotations on this function? Testing locally TS knows what it is without it.

Reply:
i think we should leave it. while building this out w/o it you can easily invalidate this and then it gets very confusing. this way you'll get an error that's easy to debug